addons/stock/models/stock_move_line.py

```diff
@@ -369,6 +370,10 @@ def _action_done(self):
         is done.
         """
         Quant = self.env['stock.quant']
+        # Assign an owner on move line if the move has a restrict partner
+        for move_line in self:
+            if not move_line.owner_id and move_line.move_id.restrict_partner_id:
+                move_line.owner_id = move_line.move_id.restrict_partner_id
```

Comment:
you could integrate this change in the loop below, or in button_validate of the picking directly